src/UnionSums.hs

```diff
@@ -74,3 +86,49 @@ stripSuffix suf = fmap unpack . Text.stripSuffix (pack suf) . pack
 
 lowerFirst :: String -> String
 lowerFirst (c:s) = toLower c : s
+
+replace :: Int -> a -> [a] -> [a]
+replace i a as = go 0 as
+  where
+    go _ [] = []
+    go currI (x:xs) =
+      let val = if currI == i then a else x in
+      val : go (currI + 1) xs
+
+-- | Produces a conversion function akin to `either` for converting "bigger"
+--   types into some other type via a value of their "smaller" type. E.g.
+--
+--   mkDecompose ''BigType [''SubTypeA, ''SubTypB]
+--
+--   defines
+--
+--   decomposeBigType :: (SubTypeA -> a) -> (SubTypeB -> a) -> BigType -> a
+--   decomposeBigType f _ (FooBigType i) = f (FooSubTypeA i)
+--   decomposeBigType f _ (BarBigType c) = f (BarSubTypeA c)
+--   decomposeBigType _ g BazBigType = g BazSubTypeB
+mkDecompose :: Name -> [Name] -> Q [Dec]
+mkDecompose unionName subNames = do
+    -- FIXME: this should probably only be called by our code, because otherwise
+    -- users could scuff up which sub-types they include for the main type.
+    unionCons <- getConstructors unionName
+    subConss <- mapM getConstructors subNames
+    let subConIdxs = mconcat $ zipWith replicate (length <$> subConss) [0..]
+    clauses <- zipWithM mkClause unionCons $ zip subConIdxs (mconcat subConss)
+    return [SigD name ft, FunD name clauses]
+  where
+    name = mkName $ "decompose" ++ nameBase unionName
+    mkArrow t1 t2 = AppT (AppT ArrowT t1) t2
+    a = VarT $ mkName "a"
+    arrows = flip mkArrow a <$> ConT <$> subNames
+    ft = foldr mkArrow (mkArrow (ConT unionName) a) arrows
+    mkClause :: Con -> (Int, Con) -> Q Clause
+    mkClause (NormalC conNameU argsU) (i, (NormalC conNameS argsS)) = do
+        -- We assume argsU and argsS have been lined up correctly!
+        argNames <- mapM (const $ newName "a") argsU
+        _Pats <- mapM (const $ return WildP) subNames
```

Comment:
Not entirely sure what the WildP is getting at, I know it is a wildcard in a pattern match, and one indexed variable is being filled in, but I couldn't say why.